File: app/src/main/java/com/example/lcsc_android_erp/core/ble/smart/SmartChassisCodec.kt
package com.example.lcsc_android_erp.core.ble.smart

import java.nio.charset.StandardCharsets
import java.util.Locale

object SmartChassisCodec {
    fun parseAdvertisement(manufacturerData: ByteArray): SmartChassisAdvertisement? {
        if (manufacturerData.size < SmartChassisProtocol.ADVERTISEMENT_CORE_SIZE) {
            return null
        }
        return SmartChassisAdvertisement(
            companyId = manufacturerData.u16Le(0),
            protoVersion = manufacturerData.u8(2),
            batchId = manufacturerData.u16Le(3),
            batteryPct = manufacturerData.u8(5),
            statusFlags = manufacturerData.u8(6),
            tableSeqLow16 = manufacturerData.u16Le(7)
        )
    }

    fun parseAndroidManufacturerPayload(
        companyId: Int,
        payload: ByteArray
    ): SmartChassisAdvertisement? {
        if (payload.size < SmartChassisProtocol.ANDROID_MANUFACTURER_PAYLOAD_CORE_SIZE) {
            return null
        }
        return SmartChassisAdvertisement(
            companyId = companyId,
            protoVersion = payload.u8(0),
            batchId = payload.u16Le(1),
            batteryPct = payload.u8(3),
            statusFlags = payload.u8(4),
            tableSeqLow16 = payload.u16Le(5)
        )
    }

    fun encodeSlotRecord(
        slot: Int,
        partId: String,
        quantity: Int,
        flags: Int
    ): ByteArray {
        require(slot in 1..SmartChassisProtocol.SLOT_COUNT) { "slot must be 1..25" }
        require(quantity in 0..0xFFFF) { "quantity must fit uint16" }
        require(flags in 0..0xFF) { "flags must fit uint8" }
        val normalizedPartId = partId.trim().uppercase(Locale.ROOT)
        require(normalizedPartId.length <= 10) { "part_id must be at most 10 ASCII bytes" }
        require(normalizedPartId.all { it.code in 0x21..0x7E }) { "part_id must be printable ASCII" }

        val bytes = ByteArray(SmartChassisProtocol.SLOT_RECORD_SIZE)
        bytes[0] = slot.toByte()
        val partBytes = normalizedPartId.toByteArray(StandardCharsets.US_ASCII)
        partBytes.copyInto(bytes, destinationOffset = 1, endIndex = partBytes.size)
        bytes.putU16Le(11, quantity)
        bytes[13] = flags.toByte()
        bytes[14] = 0
        bytes[15] = crc8Maxim(bytes, 0, SmartChassisProtocol.SLOT_RECORD_SIZE - 1).toByte()
        return bytes
    }

    fun parseSlotRecord(bytes: ByteArray, requireValidCrc: Boolean = true): SmartChassisSlotRecord? {
        if (bytes.size != SmartChassisProtocol.SLOT_RECORD_SIZE) {
            return null
        }
        val crc = bytes.u8(15)
        val computed = crc8Maxim(bytes, 0, SmartChassisProtocol.SLOT_RECORD_SIZE - 1)
        if (requireValidCrc && crc != computed) {
            return null
        }
        return SmartChassisSlotRecord(
            slot = bytes.u8(0),
            partId = bytes.decodePartId(),
            quantity = bytes.u16Le(11),
            flags = bytes.u8(13),
            crc8 = crc
        )
    }

    fun encodeReadOne(slot: Int): ByteArray {
        require(slot in 1..SmartChassisProtocol.SLOT_COUNT) { "slot must be 1..25" }
        return byteArrayOf(SmartChassisBindingOp.READ_ONE.code.toByte(), slot.toByte())
    }

    fun encodeReadAll(): ByteArray {
        return byteArrayOf(SmartChassisBindingOp.READ_ALL.code.toByte())
    }

    fun encodeWriteOne(record: ByteArray): ByteArray {
        require(record.size == SmartChassisProtocol.SLOT_RECORD_SIZE) { "record must be 16 bytes" }
        require(record.u8(0) in 1..SmartChassisProtocol.SLOT_COUNT) { "record slot must be 1..25" }
        return byteArrayOf(SmartChassisBindingOp.WRITE_ONE.code.toByte()) + record
    }

    fun encodeClearOne(slot: Int): ByteArray {
        require(slot in 1..SmartChassisProtocol.SLOT_COUNT) { "slot must be 1..25" }
        return byteArrayOf(SmartChassisBindingOp.CLEAR_ONE.code.toByte(), slot.toByte())
    }

    fun encodeInsertAt(slot: Int, record: ByteArray): ByteArray {
        require(slot in 1..SmartChassisProtocol.SLOT_COUNT) { "slot must be 1..25" }
        require(record.size == SmartChassisProtocol.SLOT_RECORD_SIZE) { "record must be 16 bytes" }
        require(record.u8(0) in 1..SmartChassisProtocol.SLOT_COUNT) { "record slot must be 1..25" }
        return byteArrayOf(SmartChassisBindingOp.INSERT_AT.code.toByte(), slot.toByte()) + record
    }

    fun encodeRemoveAt(slot: Int): ByteArray {
        require(slot in 1..SmartChassisProtocol.SLOT_COUNT) { "slot must be 1..25" }
        return byteArrayOf(SmartChassisBindingOp.REMOVE_AT.code.toByte(), slot.toByte())
    }

    fun encodeMoveBlock(from: Int, to: Int, length: Int): ByteArray {
        require(from in 1..SmartChassisProtocol.SLOT_COUNT) { "from must be 1..25" }
        require(to in 1..SmartChassisProtocol.SLOT_COUNT) { "to must be 1..25" }
        require(length in 1..SmartChassisProtocol.SLOT_COUNT) { "length must be 1..25" }
        require(from + length - 1 <= SmartChassisProtocol.SLOT_COUNT) { "from + length must stay within 25 slots" }
        require(to + length - 1 <= SmartChassisProtocol.SLOT_COUNT) { "to + length must stay within 25 slots" }
        return byteArrayOf(
            SmartChassisBindingOp.MOVE_BLOCK.code.toByte(),
            from.toByte(),
            to.toByte(),
            length.toByte()
        )
    }

    fun encodeSetQuantity(slot: Int, quantity: Int): ByteArray {
        require(slot in 1..SmartChassisProtocol.SLOT_COUNT) { "slot must be 1..25" }
        require(quantity in 0..0xFFFF) { "quantity must fit uint16" }
        return ByteArray(4).also { bytes ->
            bytes[0] = SmartChassisBindingOp.SET_QTY.code.toByte()
            bytes[1] = slot.toByte()
            bytes.putU16Le(2, quantity)
        }
    }

    fun encodeFactoryReset(): ByteArray {
        return ByteArray(5).also { bytes ->
            bytes[0] = SmartChassisBindingOp.FACTORY_RESET.code.toByte()
            bytes.putU32Le(1, SmartChassisProtocol.FACTORY_RESET_MAGIC)
        }
    }

    fun parseBindingResult(bytes: ByteArray): SmartChassisBindingResult? {
        if (bytes.size < 2) {
            return null
        }
        val rawOp = bytes.u8(0)
        val rawStatus = bytes.u8(1)
        return SmartChassisBindingResult(
            op = SmartChassisBindingOp.fromCode(rawOp),
            rawOp = rawOp,
            status = SmartChassisBindingStatus.fromCode(rawStatus),
            rawStatus = rawStatus,
            payload = bytes.copyOfRange(2, bytes.size)
        )
    }

    fun isReadAllEndPayload(payload: ByteArray): Boolean {
        return payload.size == 1 && payload.u8(0) == SmartChassisProtocol.READ_ALL_END_MARKER
    }

    fun parseTableInfo(bytes: ByteArray): SmartChassisTableInfo? {
        if (bytes.size != SmartChassisProtocol.TABLE_INFO_SIZE) {
            return null
        }
        return SmartChassisTableInfo(
            tableSeq = bytes.u32Le(0),
            crc16 = bytes.u16Le(4),
            slotCount = bytes.u8(6)
        )
    }

    fun encodeLightCommand(command: SmartChassisLightCommand): ByteArray {
        require(command.mode != SmartChassisLightMode.UNKNOWN) { "mode must be a known light mode" }
        require(command.maskA in 0..SmartChassisProtocol.SLOT_MASK_MAX) { "maskA only supports 25 slots" }
        require(command.maskB in 0..SmartChassisProtocol.SLOT_MASK_MAX) { "maskB only supports 25 slots" }
        require(command.timeoutSeconds in 0..SmartChassisProtocol.MAX_LIGHT_TIMEOUT_SECONDS) {
            "timeoutSeconds must be 0..300"
        }
        require(command.mode != SmartChassisLightMode.FX ||
            command.timeoutSeconds in 0..SmartChassisProtocol.MAX_FX_TIMEOUT_SECONDS) {
            "FX timeoutSeconds must be 0..10"
        }
        return ByteArray(SmartChassisProtocol.LIGHT_COMMAND_SIZE).also { bytes ->
            bytes[0] = command.mode.code.toByte()
            bytes.putU32Le(1, command.maskA.toLong())
            bytes.putU32Le(5, command.maskB.toLong())
            bytes.putRgb(9, command.colorA)
            bytes.putRgb(12, command.colorB)
            bytes.putU16Le(15, command.timeoutSeconds)
        }
    }

    fun parseLightStatus(bytes: ByteArray): SmartChassisLightStatus? {
        if (bytes.size != 3) {
            return null
        }
        val rawMode = bytes.u8(0)
        return SmartChassisLightStatus(
            mode = SmartChassisLightMode.fromCode(rawMode),
            rawMode = rawMode,
            remainingSeconds = bytes.u16Le(1)
        )
    }

    fun slotMask(slot: Int): Int {
        return if (slot in 1..SmartChassisProtocol.SLOT_COUNT) {
            1 shl (slot - 1)
        } else {
            0
        }
    }

    fun crc8Maxim(bytes: ByteArray, offset: Int = 0, length: Int = bytes.size): Int {
        var crc = 0x00
        for (index in offset until offset + length) {
            crc = crc xor bytes.u8(index)
            repeat(8) {
                crc = if ((crc and 0x01) != 0) {
                    (crc ushr 1) xor 0x8C
                } else {
                    crc ushr 1
                }
            }
        }
        return crc and 0xFF
    }

    fun crc16CcittFalse(bytes: ByteArray, offset: Int = 0, length: Int = bytes.size): Int {
        var crc = 0xFFFF
        for (index in offset until offset + length) {
            crc = crc xor (bytes.u8(index) shl 8)
            repeat(8) {
                crc = if ((crc and 0x8000) != 0) {
                    (crc shl 1) xor 0x1021
                } else {
                    crc shl 1
                }
                crc = crc and 0xFFFF
            }
        }
        return crc and 0xFFFF
    }

    private fun ByteArray.decodePartId(): String {
        val length = (1 until 11).firstOrNull { index -> this[index].toInt() == 0 }?.minus(1) ?: 10
        return String(this, 1, length, StandardCharsets.US_ASCII).trim()
    }

    private fun ByteArray.u8(offset: Int): Int = this[offset].toInt() and 0xFF

    private fun ByteArray.u16Le(offset: Int): Int {
        return u8(offset) or (u8(offset + 1) shl 8)
    }

    private fun ByteArray.u32Le(offset: Int): Long {
        return u8(offset).toLong() or
            (u8(offset + 1).toLong() shl 8) or
            (u8(offset + 2).toLong() shl 16) or
            (u8(offset + 3).toLong() shl 24)
    }

    private fun ByteArray.putU16Le(offset: Int, value: Int) {
        this[offset] = (value and 0xFF).toByte()
        this[offset + 1] = ((value ushr 8) and 0xFF).toByte()
    }

    private fun ByteArray.putU32Le(offset: Int, value: Long) {
        this[offset] = (value and 0xFF).toByte()
        this[offset + 1] = ((value ushr 8) and 0xFF).toByte()
        this[offset + 2] = ((value ushr 16) and 0xFF).toByte()
        this[offset + 3] = ((value ushr 24) and 0xFF).toByte()
    }

    private fun ByteArray.putRgb(offset: Int, color: RgbColor) {
        require(color.red in 0..0xFF && color.green in 0..0xFF && color.blue in 0..0xFF) {
            "RGB channels must fit uint8"
        }
        this[offset] = color.red.toByte()
        this[offset + 1] = color.green.toByte()
        this[offset + 2] = color.blue.toByte()
    }
}

File: ios/VibeRack/Core/Protocol/SmartChassisCodec.swift
import Foundation

public enum SmartChassisCodec {
    public static func parseAdvertisement(_ data: Data) -> SmartChassisAdvertisement? {
        guard data.count >= SmartChassisProtocol.advertisementCoreSize else {
            return nil
        }

        return SmartChassisAdvertisement(
            companyId: data.u16Le(at: 0),
            protoVersion: data.u8(at: 2),
            batchId: data.u16Le(at: 3),
            batteryPct: data.u8(at: 5),
            statusFlags: data.u8(at: 6),
            tableSeqLow16: data.u16Le(at: 7)
        )
    }

    public static func parseAndroidManufacturerPayload(
        companyId: UInt16,
        payload: Data
    ) -> SmartChassisAdvertisement? {
        guard payload.count >= SmartChassisProtocol.androidManufacturerPayloadCoreSize else {
            return nil
        }

        return SmartChassisAdvertisement(
            companyId: companyId,
            protoVersion: payload.u8(at: 0),
            batchId: payload.u16Le(at: 1),
            batteryPct: payload.u8(at: 3),
            statusFlags: payload.u8(at: 4),
            tableSeqLow16: payload.u16Le(at: 5)
        )
    }

    public static func encodeSlotRecord(
        slot: Int,
        partId: String,
        quantity: Int,
        flags: Int
    ) -> Data {
        precondition(1...SmartChassisProtocol.slotCount ~= slot, "slot must be 1..25")
        precondition(0...0xFFFF ~= quantity, "quantity must fit uint16")
        precondition(0...0xFF ~= flags, "flags must fit uint8")

        let normalizedPartId = partId.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard let partBytes = normalizedPartId.data(using: .ascii) else {
            preconditionFailure("part_id must be printable ASCII")
        }
        precondition(!partBytes.isEmpty, "part_id must not be empty")
        precondition(partBytes.count <= 10, "part_id must be at most 10 ASCII bytes")
        precondition(partBytes.allSatisfy { 0x21...0x7E ~= $0 }, "part_id must be printable ASCII")

        var bytes = Data(repeating: 0, count: SmartChassisProtocol.slotRecordSize)
        bytes[0] = UInt8(slot)
        bytes.replaceSubrange(1..<(1 + partBytes.count), with: partBytes)
        bytes.putU16Le(UInt16(quantity), at: 11)
        bytes[13] = UInt8(flags)
        bytes[14] = 0
        bytes[15] = crc8Maxim(bytes, offset: 0, length: SmartChassisProtocol.slotRecordSize - 1)
        return bytes
    }

    public static func parseSlotRecord(
        _ data: Data,
        requireValidCRC: Bool = true
    ) -> SlotRecord? {
        guard data.count == SmartChassisProtocol.slotRecordSize else {
            return nil
        }

        let crc = data.u8(at: 15)
        let computed = crc8Maxim(data, offset: 0, length: SmartChassisProtocol.slotRecordSize - 1)
        if requireValidCRC && crc != computed {
            return nil
        }

        let slot = Int(data.u8(at: 0))
        guard slot <= SmartChassisProtocol.slotCount else {
            return nil
        }

        return SlotRecord(
            slot: slot,
            partId: data.decodePartId(),
            quantity: Int(data.u16Le(at: 11)),
            flags: data.u8(at: 13),
            crc8: crc
        )
    }

    public static func encodeReadOne(slot: Int) -> Data {
        precondition(1...SmartChassisProtocol.slotCount ~= slot, "slot must be 1..25")
        return Data([BindingOp.readOne.code, UInt8(slot)])
    }

    public static func encodeReadAll() -> Data {
        Data([BindingOp.readAll.code])
    }

    public static func encodeWriteOne(record: Data) -> Data {
        precondition(record.count == SmartChassisProtocol.slotRecordSize, "record must be 16 bytes")
        precondition(1...SmartChassisProtocol.slotCount ~= Int(record.u8(at: 0)), "record slot must be 1..25")
        return Data([BindingOp.writeOne.code]) + record
    }

    public static func encodeClearOne(slot: Int) -> Data {
        precondition(1...SmartChassisProtocol.slotCount ~= slot, "slot must be 1..25")
        return Data([BindingOp.clearOne.code, UInt8(slot)])
    }

    public static func encodeInsertAt(slot: Int, record: Data) -> Data {
        precondition(1...SmartChassisProtocol.slotCount ~= slot, "slot must be 1..25")
        precondition(record.count == SmartChassisProtocol.slotRecordSize, "record must be 16 bytes")
        precondition(1...SmartChassisProtocol.slotCount ~= Int(record.u8(at: 0)), "record slot must be 1..25")
        return Data([BindingOp.insertAt.code, UInt8(slot)]) + record
    }

    public static func encodeRemoveAt(slot: Int) -> Data {
        precondition(1...SmartChassisProtocol.slotCount ~= slot, "slot must be 1..25")
        return Data([BindingOp.removeAt.code, UInt8(slot)])
    }

    public static func encodeMoveBlock(from: Int, to: Int, length: Int) -> Data {
        precondition(1...SmartChassisProtocol.slotCount ~= from, "from must be 1..25")
        precondition(1...SmartChassisProtocol.slotCount ~= to, "to must be 1..25")
        precondition(1...SmartChassisProtocol.slotCount ~= length, "length must be 1..25")
        precondition(from + length - 1 <= SmartChassisProtocol.slotCount, "from + length must stay within 25 slots")
        precondition(to + length - 1 <= SmartChassisProtocol.slotCount, "to + length must stay within 25 slots")
        return Data([BindingOp.moveBlock.code, UInt8(from), UInt8(to), UInt8(length)])
    }

    public static func encodeSetQuantity(slot: Int, quantity: Int) -> Data {
        precondition(1...SmartChassisProtocol.slotCount ~= slot, "slot must be 1..25")
        precondition(0...0xFFFF ~= quantity, "quantity must fit uint16")

        var bytes = Data(repeating: 0, count: 4)
        bytes[0] = BindingOp.setQuantity.code
        bytes[1] = UInt8(slot)
        bytes.putU16Le(UInt16(quantity), at: 2)
        return bytes
    }

    public static func encodeFactoryReset() -> Data {
        var bytes = Data(repeating: 0, count: 5)
        bytes[0] = BindingOp.factoryReset.code
        bytes.putU32Le(SmartChassisProtocol.factoryResetMagic, at: 1)
        return bytes
    }

    public static func parseBindingResult(_ data: Data) -> BindingResult? {
        guard data.count >= 2 else {
            return nil
        }

        let rawOp = data.u8(at: 0)
        let rawStatus = data.u8(at: 1)
        return BindingResult(
            op: BindingOp(rawValue: rawOp),
            rawOp: rawOp,
            status: BindingStatus(rawValue: rawStatus),
            rawStatus: rawStatus,
            payload: data.dropFirst(2).data
        )
    }

    public static func isReadAllEndPayload(_ data: Data) -> Bool {
        data.count == 1 && data.u8(at: 0) == SmartChassisProtocol.readAllEndMarker
    }

    public static func parseTableInfo(_ data: Data) -> TableInfo? {
        guard data.count == SmartChassisProtocol.tableInfoSize else {
            return nil
        }

        return TableInfo(
            tableSeq: data.u32Le(at: 0),
            crc16: data.u16Le(at: 4),
            slotCount: Int(data.u8(at: 6))
        )
    }

    public static func encodeLightCommand(_ command: LightCommand) -> Data {
        precondition(!command.mode.isUnknown, "mode must be a known light mode")
        precondition(command.maskA <= SmartChassisProtocol.slotMaskMax, "maskA only supports 25 slots")
        precondition(command.maskB <= SmartChassisProtocol.slotMaskMax, "maskB only supports 25 slots")
        precondition(
            0...SmartChassisProtocol.maxLightTimeoutSeconds ~= command.timeoutSeconds,
            "timeoutSeconds must be 0..300"
        )
        precondition(
            command.mode != .fx || 0...SmartChassisProtocol.maxFxTimeoutSeconds ~= command.timeoutSeconds,
            "FX timeoutSeconds must be 0..10"
        )

        var bytes = Data(repeating: 0, count: SmartChassisProtocol.lightCommandSize)
        bytes[0] = command.mode.code
        bytes.putU32Le(command.maskA, at: 1)
        bytes.putU32Le(command.maskB, at: 5)
        bytes.putRGB(command.colorA, at: 9)
        bytes.putRGB(command.colorB, at: 12)
        bytes.putU16Le(UInt16(command.timeoutSeconds), at: 15)
        return bytes
    }

    public static func parseLightStatus(_ data: Data) -> LightStatus? {
        guard data.count == 3 else {
            return nil
        }

        let rawMode = data.u8(at: 0)
        return LightStatus(
            mode: LightMode(rawValue: rawMode),
            rawMode: rawMode,
            remainingSeconds: Int(data.u16Le(at: 1))
        )
    }

    public static func slotMask(slot: Int) -> UInt32 {
        guard 1...SmartChassisProtocol.slotCount ~= slot else {
            return 0
        }

        return 1 << UInt32(slot - 1)
    }

    public static func crc8Maxim(
        _ data: Data,
        offset: Int = 0,
        length: Int? = nil
    ) -> UInt8 {
        let byteCount = length ?? data.count
        precondition(offset >= 0 && byteCount >= 0 && offset + byteCount <= data.count, "CRC range out of bounds")

        var crc: UInt8 = 0x00
        for index in offset..<(offset + byteCount) {
            crc ^= data.u8(at: index)
            for _ in 0..<8 {
                if crc & 0x01 != 0 {
                    crc = (crc >> 1) ^ 0x8C
                } else {
                    crc >>= 1
                }
            }
        }
        return crc
    }

    public static func crc16CcittFalse(
        _ data: Data,
        offset: Int = 0,
        length: Int? = nil
    ) -> UInt16 {
        let byteCount = length ?? data.count
        precondition(offset >= 0 && byteCount >= 0 && offset + byteCount <= data.count, "CRC range out of bounds")

        var crc: UInt16 = 0xFFFF
        for index in offset..<(offset + byteCount) {
            crc ^= UInt16(data.u8(at: index)) << 8
            for _ in 0..<8 {
                if crc & 0x8000 != 0 {
                    crc = (crc << 1) ^ 0x1021
                } else {
                    crc <<= 1
                }
            }
        }
        return crc
    }

    public static func parseDeviceHealth(_ data: Data) -> DeviceHealth? {
        guard data.count == 4 else {
            return nil
        }

        return DeviceHealth(
            batteryPct: data.u8(at: 0),
            resetReason: data.u16Le(at: 1),
            healthFlags: data.u8(at: 3)
        )
    }
}

private extension Data.SubSequence {
    var data: Data {
        Data(self)
    }
}

private extension LightMode {
    var isUnknown: Bool {
        if case .unknown = self {
            return true
        }
        return false
    }
}

private extension Data {
    func u8(at offset: Int) -> UInt8 {
        self[startIndex + offset]
    }

    func u16Le(at offset: Int) -> UInt16 {
        UInt16(u8(at: offset)) | (UInt16(u8(at: offset + 1)) << 8)
    }

    func u32Le(at offset: Int) -> UInt32 {
        UInt32(u8(at: offset)) |
            (UInt32(u8(at: offset + 1)) << 8) |
            (UInt32(u8(at: offset + 2)) << 16) |
            (UInt32(u8(at: offset + 3)) << 24)
    }

    mutating func putU16Le(_ value: UInt16, at offset: Int) {
        self[startIndex + offset] = UInt8(value & 0x00FF)
        self[startIndex + offset + 1] = UInt8((value >> 8) & 0x00FF)
    }

    mutating func putU32Le(_ value: UInt32, at offset: Int) {
        self[startIndex + offset] = UInt8(value & 0x0000_00FF)
        self[startIndex + offset + 1] = UInt8((value >> 8) & 0x0000_00FF)
        self[startIndex + offset + 2] = UInt8((value >> 16) & 0x0000_00FF)
        self[startIndex + offset + 3] = UInt8((value >> 24) & 0x0000_00FF)
    }

    mutating func putRGB(_ color: RGBColor, at offset: Int) {
        self[startIndex + offset] = color.red
        self[startIndex + offset + 1] = color.green
        self[startIndex + offset + 2] = color.blue
    }

    func decodePartId() -> String {
        let partBytes = self[(startIndex + 1)..<(startIndex + 11)]
        let endIndex = partBytes.firstIndex(of: 0) ?? partBytes.endIndex
        return String(data: partBytes[..<endIndex], encoding: .ascii)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    }
}
